lib/ws_deque.ml

```diff
@@ -0,0 +1,196 @@
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+(* Work Stealing Queue
+ *
+ * See:
+ *   Dynamic circular work-stealing deque
+ *   https://dl.acm.org/doi/10.1145/1073970.1073974
+ *  &
+ *   Correct and efficient work-stealing for weak memory models
+ *   https://dl.acm.org/doi/abs/10.1145/2442516.2442524
+ *)
+
+module type S = sig
+  type 'a t
+  val create : unit -> 'a t
+  val is_empty : 'a t -> bool
+  val size : 'a t -> int
+  val push : 'a t -> 'a -> unit
+  val pop : 'a t -> 'a
+  val steal : 'a t -> 'a
+end
+
+module CArray = struct
+
+  type 'a t = {
+    arr  : 'a array;
+    mask : int
+    }
+
+  let rec log2 n =
+    if n <= 1 then 0 else 1 + (log2 (n asr 1))
+
+  let create n v =
+    let sz = Int.shift_left 1 (log2 n) in
+    assert ((sz >= n) && (sz > 0));
+    assert (Int.logand sz (sz-1) == 0);
+    {
+      arr  = Array.init sz (fun _ -> v);
+      mask = sz - 1
+    }
+
+  let size t = Array.length t.arr [@@inline]
+
+  let get t i =
+    Array.unsafe_get t.arr (Int.logand i t.mask) [@@inline]
+
+  let put t i v =
+    Array.unsafe_set t.arr (Int.logand i t.mask) v [@@inline]
+
+  let grow t top bottom =
+    let s = size t in
+    let ns = 2 * s in
+    let out = create ns (Obj.magic ()) in
+    for i = top to bottom do
+      put out i (get t i)
+    done;
+    out
+
+  let shrink t top bottom =
+    let s = size t in
+    let ns = s / 2 in
+    let out = create ns (Obj.magic ()) in
+    for i = top to bottom do
+      put out i (get t i)
+    done;
+    out
+
+end
+
+module M : S = struct
+  let min_size = 32
+  let shrink_const = 3
+
+  type 'a t = {
+    top : int Atomic.t;
+    bottom : int Atomic.t;
+    tab : 'a ref CArray.t Atomic.t;
+    mutable next_shrink : int;
+  }
+
+  let create () = {
+    top = Atomic.make 1;
+    bottom = Atomic.make 1;
+    tab = Atomic.make (CArray.create min_size (Obj.magic ()));
+    next_shrink = 0
+  }
+
+  let set_next_shrink q =
+    let sz = CArray.size (Atomic.get q.tab) in
+    if sz <= min_size then
+      q.next_shrink <- 0
+    else
+      q.next_shrink <- sz / shrink_const
+
+  let grow q t b =
+    Atomic.set q.tab (CArray.grow (Atomic.get q.tab) t b);
+    set_next_shrink q
+
+  let is_empty q =
+    let b = Atomic.get q.bottom in
+    let t = Atomic.get q.top in
+    b - t <= 0
+
+  let size q =
+    let b = Atomic.get q.bottom in
+    let t = Atomic.get q.top in
+    b - t
+
+  let push q v =
+    let v' = ref v in
+    let b = Atomic.get q.bottom in
+    let t = Atomic.get q.top in
+    let a = Atomic.get q.tab in
+    let size = b - t in
+    let a =
+      if size >= CArray.size a - 1 then
```

Comment:
Can `size` ever be greater than `CArray.size a - 1`?

Reply:
Can I make this check `size == CArray.size a`? Why is the `-1` necessary?

Reply:
Good spot, this is a place where we are differing needlessly from the reference implementation (see Fig 1 of [here](https://dl.acm.org/doi/abs/10.1145/2442516.2442524)).

I will make the change.